base64ct: support both unpadded and padded encodings

tarcieri opened the pull request:
We need support for both:

- "PEM" (as used by PKCS#8) needs padded Base64
- "B64" (as used by the PHC string format) and JWK need unpadded Base64

newpavlov commented on base64ct/src/base64.rs:
```diff
@@ -0,0 +1,229 @@
+
+    while src_length >= 3 {
+        encode_3bytes(
+            &src[src_offset..(src_offset + 3)],
```
Why use the `while` loop instead of the approach used in the `b64` module? The latter is more idiomatic and allows compiler to remove bound checks. Same for `decode`.

tarcieri replied:
I tried to get this working but had a little bit of trouble. The logic is a bit different from the unpadded case.

tarcieri replied:
I managed to get it working with `chunks` instead of a loop, but I couldn't get it working with `chunks_exact`. Perhaps you could take another look.

newpavlov commented on base64ct/src/base64.rs:
```diff
@@ -0,0 +1,229 @@
+
+/// Encoded len
+pub fn encoded_len(bytes: &[u8]) -> usize {
+    (((bytes.len() * 4) / 3) + 3) & !3
```
Remember the overflow bug.

tarcieri replied:
Fixed

newpavlov commented on base64ct/src/base64.rs:
```diff
@@ -0,0 +1,229 @@
+    let src = src.as_ref();
+
+    // TODO(tarcieri): constant-time whitespace tolerance
+    if !src.is_empty() && char::from(src[src.len() - 1]).is_whitespace() {
```
Are you sure you want to be so generic instead of simply checking for `b' '`?

tarcieri replied:
It should definitely handle newlines as well. They're probably the most common issue here, actually.

All that said, unless there's a separate error variant for whitespace versus other characters, having an eager whitespace check probably provides little value. So we should either add another error variant, or just remove this check entirely.

newpavlov commented on base64ct/src/base64.rs:
```diff
@@ -0,0 +1,229 @@
+// 0x41-0x5a, 0x61-0x7a, 0x30-0x39, 0x2b, 0x2f
+
+#[inline(always)]
+pub(crate) fn encode_3bytes(src: &[u8], dst: &mut [u8]) {
```
Maybe it's worth to move these functions into a separate `utils` module?

tarcieri replied:
I wound up extracting them into a `common` module. There's probably more code that can be DRYed out between them.

newpavlov commented on base64ct/Cargo.toml:
```diff
@@ -14,9 +14,13 @@ categories = ["cryptography", "encoding", "no-std"]
 
+[dependencies]
+zeroize = ""
+
```
Shouldn't zeroization be a concern of a higher level? It's a bit strange to have it in a parsing library.

tarcieri replied:
Removed `zeroize`

tarcieri commented:
@newpavlov the short answer to all of your questions is the code in the `b64ct` crate was derived from a constant-time implementation of `=`-padded Base64, which is what I've used here.

The formulas are slightly different, and as such they do reintroduce previously solved overflow bugs, but that's for "B64" instead of padded Base64.

Also note: work-in-progress 😉

tarcieri commented:
I pushed up a slightly different version that only tries to handle the `padded` vs `unpadded` distinction for now.

Still a lot of duplication and unaddressed issues.

tarcieri commented:
Marking as ready for review.

It's still not great, but the tests are passing and I managed to boil the code down to a single module again.

There's still a lot of repetitiveness, but I ran out of gas trying to DRY it out.

I also didn't try to tackle `decode_in_place` and just renamed it to `decode_in_place_unpadded` for now.

@newpavlov perhaps we can merge this and then you can take a shot at refactoring it?

tarcieri commented:
Going to merge this and work on Base64uri support
